pysindy/optimizers/trapping_sr3.py

```diff
@@ -310,14 +441,47 @@ def _reduce(self, x, y):
         """
 
         n_samples, n_features = x.shape
-        r = (self.PL).shape[0]
-        N = n_features
+        r = y.shape[1]
+        N = int((r ** 2 + 3 * r) / 2.0)
+
+        # If these tensors are not passed, or incorrect shape, assume zeros
+        if self.PQ is None:
+            self.PQ = np.zeros((r, r, r, r, n_features))
+        elif (self.PQ).shape != (r, r, r, r, n_features) and (self.PQ).shape != (
+            r,
+            r,
+            r,
+            r,
+            N,
+        ):
+            self.PQ = np.zeros((r, r, r, r, n_features))
+        if self.PL is None:
+            self.PL = np.zeros((r, r, r, n_features))
+        elif (self.PL).shape != (r, r, r, n_features) and (self.PL).shape != (
+            r,
+            r,
+            r,
+            N,
+        ):
+            self.PL = np.zeros((r, r, r, n_features))
+
+        # Check if the tensor symmetries are properly defined
+        if self._bad_PL(self.PL):
+            raise ValueError("PL tensor was passed but not properly")
```

Comment:
Could you make these error messages a bit more descriptive?

Reply:
And what do you think about moving these checks to the `__init__` method, so that `PL` and `PQ` can be checked when they are passed in?

Reply:
I can make these more descriptive. I tried checking when they are passed but I couldn't figure out how to do this. The problem is that PL and PQ can be omitted (if the trapping theorem not being used), in which case no checks are necessary. But if PL and PQ are passed, I need to check that 1) they are of the correct dimensions are 2) The satisfy the correct symmetry properties. To check if they are the correct dimensions, I must know the shape of x and y passed to the fit (or reduce) functions, meaning this check cannot be done in the initialization

Reply:
Ignore my previous comment. It looks like you need to know the number of features to potentially reshape `PQ` and `PL`  before checking the tensor symmetries. And you don't know the number of features until `_reduce` is called.

Reply:
cool, agreed